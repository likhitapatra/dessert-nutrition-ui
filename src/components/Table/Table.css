.container {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    background-color: antiquewhite;
    padding: 16px;
}

.button-container {
    display: flex;
    flex-wrap: wrap;
    flex: 1 0;
    justify-content: flex-end;
}

.selection {
    background-color: antiquewhite;
}