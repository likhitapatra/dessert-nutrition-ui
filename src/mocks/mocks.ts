import { GET_NUTRITION_DATA } from '../queries/queries';

export const mocks = [
  {
    request: {
      query: GET_NUTRITION_DATA,
      variables: {
        dessert: 'Oreo',
      },
    },
    result: {
      data: {
        nutritionData: { dessert: 'Oreo', nutritionInfo: {} },
      },
    },
  },
];
