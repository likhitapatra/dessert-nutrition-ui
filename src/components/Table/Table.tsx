import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import TableData from '../TableData/TableData';
import TableHeader from '../TableHeader/TableHeader';
import Button from '../Button/Button';
import { TiArrowSortedUp, TiArrowSortedDown } from 'react-icons/ti';
import './Table.css';
import { useMutation } from '@apollo/client';
import {
  DELETE_NUTRITION_DATA,
  GET_NUTRITION_DATA,
} from '../../queries/queries';

const Table = (props: any) => {
  const history = useHistory();
  const [toggle, setToggle] = useState(false);

  const [state, setState] = useState(props.data.nutritionData);

  const [checked, setChecked] = useState(
    [...Array(props.data?.nutritionData.length)].map(() => false)
  );
  const [allChecked, setAllChecked] = useState(false);
  const [selectedCount, setSelectedCount] = useState(0);
  const [selectedDessert, setSelectedDessert] = useState('');
  const [deleteNutritionData] = useMutation(DELETE_NUTRITION_DATA);

  const handleSelected = (event: any, currentIndex: any) => {
    const _checked = [...checked];
    _checked[currentIndex] = event.target.checked;
    setChecked(_checked);
    event.target.checked
      ? setSelectedDessert(props.data.nutritionData[currentIndex].dessert)
      : setSelectedDessert('');

    const checkedCount = _checked.filter((val) => val === true).length;
    setSelectedCount(checkedCount);
  };

  const handleSelectAll = (event: any) => {
    setAllChecked(event.target.checked);
    setChecked((checked) => checked.map(() => event.target.checked));
    !allChecked ? setSelectedCount(checked.length) : setSelectedCount(0);
  };

  const handleAdd = () => {
    history.push('/add');
  };

  const handleDeleteSelected = () => {
    deleteNutritionData({
      variables: { dessert: selectedDessert },
      refetchQueries: [{ query: GET_NUTRITION_DATA }],
    });
  };

  const sortAscending = (sortKey: string) => {
    setToggle(!toggle);
    const sortedData = props.data?.nutritionData;
    const result = sortedData
      .slice()
      .sort((a: any, b: any) => (a[sortKey] > b[sortKey] ? 1 : -1));

    setState(result);
  };

  const sortDescending = (sortKey: string) => {
    setToggle(!toggle);
    const sortedData = props.data?.nutritionData;
    const result = sortedData
      .slice()
      .sort((a: any, b: any) => (a[sortKey] < b[sortKey] ? 1 : -1));

    setState(result);
  };

  const tableHeaders = [
    {
      name: (
        <input
          type="checkbox"
          id="desserts"
          name="desserts"
          checked={allChecked}
          onChange={handleSelectAll}
        />
      ),
      key: 'something',
    },
    { name: 'Dessert (100g serving)', key: 'dessert' },
    { name: 'Calories', key: 'calories' },
    { name: 'Fat (g)', key: 'fat' },
    { name: 'Carbs (g)', key: 'carb' },
    { name: 'Protein (g)', key: 'protein' },
  ];

  return (
    <div>
      <div className="container">
        <div className="selection">{selectedCount} Selected</div>
        <div className="button-container">
          <div className="ph2">
            <Button color="green" onClick={handleAdd}>
              + Add New
            </Button>
          </div>
          <Button color="red" onClick={handleDeleteSelected}>
            - Delete
          </Button>
        </div>
      </div>
      <div className="pa4">
        <div className="overflow-auto">
          <table className="f6 w-100 mw12 center" cellSpacing="0">
            <thead>
              <tr>
                {tableHeaders.map(({ name, key }: any, index) => (
                  <TableHeader key={index}>
                    {toggle ? (
                      <TiArrowSortedUp onClick={() => sortAscending(key)} />
                    ) : (
                      <TiArrowSortedDown onClick={() => sortDescending(key)} />
                    )}
                    {name}
                  </TableHeader>
                ))}
              </tr>
            </thead>
            <tbody className="lh-copy">
              {state.map((values: any, index: string) => (
                <tr key={index}>
                  <TableData>
                    <input
                      type="checkbox"
                      checked={checked[parseInt(index)]}
                      onChange={(event) =>
                        handleSelected(event, parseInt(index))
                      }
                    />
                  </TableData>
                  <TableData>{values.dessert}</TableData>
                  <TableData>{values.nutritionInfo.calories}</TableData>
                  <TableData>{values.nutritionInfo.fat}</TableData>
                  <TableData>{values.nutritionInfo.carb}</TableData>
                  <TableData>{values.nutritionInfo.protein}</TableData>
                </tr>
              ))}
            </tbody>
          </table>
        </div>
      </div>
    </div>
  );
};

export default Table;
