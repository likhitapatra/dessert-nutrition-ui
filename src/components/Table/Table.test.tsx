import React from 'react';
import { MockedProvider } from '@apollo/client/testing';
import { act } from 'react-test-renderer';

import { mocks } from '../../mocks/mocks';
import Table from './Table';
import { render } from '@testing-library/react';

describe('Table', () => {
  const props = { data: { nutritionData: [] } };
  it('should match Table snapshot', async () => {
    let rendered;
    await act(async () => {
      rendered = render(
        <MockedProvider mocks={mocks}>
          <Table {...props} />
        </MockedProvider>
      );
    });
    expect(rendered).toMatchSnapshot();
  });

  it('should render Table component', async () => {
    await act(async () => {
      const test = render(
        <MockedProvider mocks={mocks}>
          <Table {...props} />
        </MockedProvider>
      );
      expect(test).toBeTruthy();
    });
  });
});
