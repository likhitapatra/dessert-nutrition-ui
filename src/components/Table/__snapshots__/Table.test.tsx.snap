// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should match Table snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="container"
        >
          <div
            class="selection"
          >
            0
             Selected
          </div>
          <div
            class="button-container"
          >
            <div
              class="ph2"
            >
              <button
                class="f6 link dim br1 ph3 pv2 mb2 dib white bg-dark-green w-undefined"
              >
                + Add New
              </button>
            </div>
            <button
              class="f6 link dim br1 ph3 pv2 mb2 dib white bg-dark-red w-undefined"
            >
              - Delete
            </button>
          </div>
        </div>
        <div
          class="pa4"
        >
          <div
            class="overflow-auto"
          >
            <table
              cellspacing="0"
              class="f6 w-100 mw12 center"
            >
              <thead>
                <tr>
                  <th
                    class="fw6 bb b--black-20 tl pb3 pr3 bg-white"
                  >
                    <svg
                      baseProfile="tiny"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      version="1.2"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M5.8 9.7l6.2 6.3 6.2-6.3c.2-.2.3-.5.3-.7s-.1-.5-.3-.7c-.2-.2-.4-.3-.7-.3h-11c-.3 0-.5.1-.7.3-.2.2-.3.4-.3.7s.1.5.3.7z"
                      />
                    </svg>
                    <input
                      id="desserts"
                      name="desserts"
                      type="checkbox"
                    />
                  </th>
                  <th
                    class="fw6 bb b--black-20 tl pb3 pr3 bg-white"
                  >
                    <svg
                      baseProfile="tiny"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      version="1.2"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M5.8 9.7l6.2 6.3 6.2-6.3c.2-.2.3-.5.3-.7s-.1-.5-.3-.7c-.2-.2-.4-.3-.7-.3h-11c-.3 0-.5.1-.7.3-.2.2-.3.4-.3.7s.1.5.3.7z"
                      />
                    </svg>
                    Dessert (100g serving)
                  </th>
                  <th
                    class="fw6 bb b--black-20 tl pb3 pr3 bg-white"
                  >
                    <svg
                      baseProfile="tiny"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      version="1.2"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M5.8 9.7l6.2 6.3 6.2-6.3c.2-.2.3-.5.3-.7s-.1-.5-.3-.7c-.2-.2-.4-.3-.7-.3h-11c-.3 0-.5.1-.7.3-.2.2-.3.4-.3.7s.1.5.3.7z"
                      />
                    </svg>
                    Calories
                  </th>
                  <th
                    class="fw6 bb b--black-20 tl pb3 pr3 bg-white"
                  >
                    <svg
                      baseProfile="tiny"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      version="1.2"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M5.8 9.7l6.2 6.3 6.2-6.3c.2-.2.3-.5.3-.7s-.1-.5-.3-.7c-.2-.2-.4-.3-.7-.3h-11c-.3 0-.5.1-.7.3-.2.2-.3.4-.3.7s.1.5.3.7z"
                      />
                    </svg>
                    Fat (g)
                  </th>
                  <th
                    class="fw6 bb b--black-20 tl pb3 pr3 bg-white"
                  >
                    <svg
                      baseProfile="tiny"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      version="1.2"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M5.8 9.7l6.2 6.3 6.2-6.3c.2-.2.3-.5.3-.7s-.1-.5-.3-.7c-.2-.2-.4-.3-.7-.3h-11c-.3 0-.5.1-.7.3-.2.2-.3.4-.3.7s.1.5.3.7z"
                      />
                    </svg>
                    Carbs (g)
                  </th>
                  <th
                    class="fw6 bb b--black-20 tl pb3 pr3 bg-white"
                  >
                    <svg
                      baseProfile="tiny"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      version="1.2"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M5.8 9.7l6.2 6.3 6.2-6.3c.2-.2.3-.5.3-.7s-.1-.5-.3-.7c-.2-.2-.4-.3-.7-.3h-11c-.3 0-.5.1-.7.3-.2.2-.3.4-.3.7s.1.5.3.7z"
                      />
                    </svg>
                    Protein (g)
                  </th>
                </tr>
              </thead>
              <tbody
                class="lh-copy"
              />
            </table>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="container"
      >
        <div
          class="selection"
        >
          0
           Selected
        </div>
        <div
          class="button-container"
        >
          <div
            class="ph2"
          >
            <button
              class="f6 link dim br1 ph3 pv2 mb2 dib white bg-dark-green w-undefined"
            >
              + Add New
            </button>
          </div>
          <button
            class="f6 link dim br1 ph3 pv2 mb2 dib white bg-dark-red w-undefined"
          >
            - Delete
          </button>
        </div>
      </div>
      <div
        class="pa4"
      >
        <div
          class="overflow-auto"
        >
          <table
            cellspacing="0"
            class="f6 w-100 mw12 center"
          >
            <thead>
              <tr>
                <th
                  class="fw6 bb b--black-20 tl pb3 pr3 bg-white"
                >
                  <svg
                    baseProfile="tiny"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    version="1.2"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M5.8 9.7l6.2 6.3 6.2-6.3c.2-.2.3-.5.3-.7s-.1-.5-.3-.7c-.2-.2-.4-.3-.7-.3h-11c-.3 0-.5.1-.7.3-.2.2-.3.4-.3.7s.1.5.3.7z"
                    />
                  </svg>
                  <input
                    id="desserts"
                    name="desserts"
                    type="checkbox"
                  />
                </th>
                <th
                  class="fw6 bb b--black-20 tl pb3 pr3 bg-white"
                >
                  <svg
                    baseProfile="tiny"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    version="1.2"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M5.8 9.7l6.2 6.3 6.2-6.3c.2-.2.3-.5.3-.7s-.1-.5-.3-.7c-.2-.2-.4-.3-.7-.3h-11c-.3 0-.5.1-.7.3-.2.2-.3.4-.3.7s.1.5.3.7z"
                    />
                  </svg>
                  Dessert (100g serving)
                </th>
                <th
                  class="fw6 bb b--black-20 tl pb3 pr3 bg-white"
                >
                  <svg
                    baseProfile="tiny"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    version="1.2"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M5.8 9.7l6.2 6.3 6.2-6.3c.2-.2.3-.5.3-.7s-.1-.5-.3-.7c-.2-.2-.4-.3-.7-.3h-11c-.3 0-.5.1-.7.3-.2.2-.3.4-.3.7s.1.5.3.7z"
                    />
                  </svg>
                  Calories
                </th>
                <th
                  class="fw6 bb b--black-20 tl pb3 pr3 bg-white"
                >
                  <svg
                    baseProfile="tiny"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    version="1.2"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M5.8 9.7l6.2 6.3 6.2-6.3c.2-.2.3-.5.3-.7s-.1-.5-.3-.7c-.2-.2-.4-.3-.7-.3h-11c-.3 0-.5.1-.7.3-.2.2-.3.4-.3.7s.1.5.3.7z"
                    />
                  </svg>
                  Fat (g)
                </th>
                <th
                  class="fw6 bb b--black-20 tl pb3 pr3 bg-white"
                >
                  <svg
                    baseProfile="tiny"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    version="1.2"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M5.8 9.7l6.2 6.3 6.2-6.3c.2-.2.3-.5.3-.7s-.1-.5-.3-.7c-.2-.2-.4-.3-.7-.3h-11c-.3 0-.5.1-.7.3-.2.2-.3.4-.3.7s.1.5.3.7z"
                    />
                  </svg>
                  Carbs (g)
                </th>
                <th
                  class="fw6 bb b--black-20 tl pb3 pr3 bg-white"
                >
                  <svg
                    baseProfile="tiny"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    version="1.2"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M5.8 9.7l6.2 6.3 6.2-6.3c.2-.2.3-.5.3-.7s-.1-.5-.3-.7c-.2-.2-.4-.3-.7-.3h-11c-.3 0-.5.1-.7.3-.2.2-.3.4-.3.7s.1.5.3.7z"
                    />
                  </svg>
                  Protein (g)
                </th>
              </tr>
            </thead>
            <tbody
              class="lh-copy"
            />
          </table>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
